pkg/regopolicyinterpreter/regopolicyinterpreter.go

```diff
@@ -0,0 +1,557 @@
+	}
+}
+
+func (r *RegoPolicyInterpreter) logMetadata() {
+	if r.logLevel < LogMetadata {
+		return
+	}
+
+	contents, err := json.Marshal(r.data["metadata"])
+	if err != nil {
+		r.metadataLogger.Printf("error marshaling metadata: %v\n", err.Error())
+	} else {
+		r.metadataLogger.Println(string(contents))
+	}
+}
+
+// Returns the raw value from a Rego query result.
+func (r RegoQueryResult) Value(key string) (interface{}, error) {
+	if value, ok := r[key]; ok {
+		return value, nil
+	} else {
+		return nil, fmt.Errorf("unable to find value for key '%s'", key)
+	}
+}
+
+// Attempts to interpret a result value as a boolean.
+func (r RegoQueryResult) Bool(key string) (bool, error) {
+	if flag, ok := r[key]; ok {
+		if value, ok := flag.(bool); ok {
+			return value, nil
+		} else {
+			return false, fmt.Errorf("value for '%s' is not a boolean", key)
+		}
+	} else {
+		return false, fmt.Errorf("unable to find value for key '%s'", key)
+	}
+}
+
+// Attempts to interpret the result value as a string.
+func (r RegoQueryResult) String(key string) (string, error) {
+	if flag, ok := r[key]; ok {
+		if value, ok := flag.(string); ok {
+			return value, nil
+		} else {
+			return "", fmt.Errorf("value for '%s' is not a string", key)
+		}
+	} else {
+		return "", fmt.Errorf("unable to find value for key '%s'", key)
+	}
+}
+
+// Attempts to interpret the result value as a floating point number.
+func (r RegoQueryResult) Float(key string) (float64, error) {
+	if value, ok := r[key]; ok {
+		if number, ok := value.(json.Number); ok {
+			if floatValue, err := number.Float64(); err == nil {
+				return floatValue, nil
+			} else {
+				return 0, err
+			}
+		} else {
+			return 0, fmt.Errorf("value for %s is not a json Number", key)
+		}
+	} else {
+		return 0, fmt.Errorf("unable to find value for key '%s'", key)
+	}
+}
+
+// Attempts to interpret the result value as an integer.
+func (r RegoQueryResult) Int(key string) (int, error) {
+	if value, ok := r[key]; ok {
+		if number, ok := value.(json.Number); ok {
+			if intValue, err := number.Int64(); err == nil {
+				return int(intValue), nil
+			} else {
+				return 0, err
+			}
+		} else {
+			return 0, errors.New("value is not a json Number")
+		}
+	} else {
+		return 0, fmt.Errorf("unable to find value for key '%s'", key)
+	}
+}
+
+// Tests if the query result is empty.
+func (r RegoQueryResult) IsEmpty() bool {
+	return len(r) == 0
+}
+
+// Queries the policy with the given rule and input data and returns the result.
+func (r *RegoPolicyInterpreter) Query(rule string, input map[string]interface{}) (RegoQueryResult, error) {
+	resultSet, output, err := r.query(rule, input)
+
+	r.logInfo(output)
+
+	if err != nil {
+		return nil, err
+	}
+
+	r.logResult(rule, resultSet)
+
+	result := make(RegoQueryResult)
+	if resultSet == nil {
+		return result, nil
+	}
+
+	ops := []*regoMetadataOperation{}
+	if metadata, ok := resultSet["metadata"].([]interface{}); ok {
```

Comment:
should we split this into two? I feel like if the metadata is not in the right format we should probably error out?
```
metadata, ok := resultSet["metadata"]
if ok {
    values, ok := metadata.([]interface{})
    if !ok {
        return nil, errors.New("metadata is invalid")
    }
}
```

Reply:
Broken out as suggested